pkg/registry/apis/provisioning/jobs/syncer.go

```diff
@@ -12,458 +12,392 @@ import (
 	"github.com/grafana/grafana/pkg/registry/apis/provisioning/resources"
-	"github.com/grafana/grafana/pkg/registry/apis/provisioning/safepath"
 )
 
-type Syncer struct {
-	client     *resources.DynamicClient
-	parser     *resources.Parser
-	lister     resources.ResourceLister
-	folders    dynamic.ResourceInterface
-	repository repository.Repository
+// Sync will make grafana look like the contents of the repository
+// it will add/remove/update resources in grafana so the results look like a mirror
+type Syncer interface {
+	Sync(ctx context.Context,
+		repo repository.Repository,
+		options provisioning.SyncJobOptions,
+		progress func(provisioning.JobStatus) error,
+	) (*provisioning.JobStatus, *provisioning.SyncStatus, error)
+}
+
+// syncer will start sync jobs
+type syncer struct {
+	parsers *resources.ParserFactory
+	lister  resources.ResourceLister
 }
 
-func NewSyncer(
+// start a job and run it
+func (r *syncer) Sync(ctx context.Context,
 	repo repository.Repository,
-	lister resources.ResourceLister,
-	parser *resources.Parser,
-) (*Syncer, error) {
+	options provisioning.SyncJobOptions,
+	progress func(provisioning.JobStatus) error,
+) (*provisioning.JobStatus, *provisioning.SyncStatus, error) {
+	parser, err := r.parsers.GetParser(ctx, repo)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get parser for %s: %w", repo.Config().Name, err)
+	}
 	dynamicClient := parser.Client()
-	folders := dynamicClient.Resource(schema.GroupVersionResource{
-		Group:    folders.GROUP,
-		Version:  folders.VERSION,
-		Resource: folders.RESOURCE,
-	})
+	if repo.Config().Namespace != dynamicClient.GetNamespace() {
+		return nil, nil, fmt.Errorf("namespace mismatch")
+	}
+	job := &syncJob{
+		repository:   repo,
+		options:      options,
+		progress:     progress,
+		parser:       parser,
+		lister:       r.lister,
+		jobStatus:    &provisioning.JobStatus{},
+		syncStatus:   &provisioning.SyncStatus{},
+		folderLookup: make(map[string]string),
+		folders: dynamicClient.Resource(schema.GroupVersionResource{
+			Group:    folders.GROUP,
+			Version:  folders.VERSION,
+			Resource: folders.RESOURCE,
+		}),
+		dashboards: dynamicClient.Resource(schema.GroupVersionResource{
+			Group:    dashboard.GROUP,
+			Version:  "v1alpha1",
+			Resource: dashboard.DASHBOARD_RESOURCE,
+		}),
+	}
+	return job.run(ctx)
+}
+
+// created once for each sync execution
+type syncJob struct {
+	repository repository.Repository
+	options    provisioning.SyncJobOptions
+	progress   func(provisioning.JobStatus) error
 
-	return &Syncer{
-		parser:     parser,
-		client:     dynamicClient,
-		lister:     lister,
-		folders:    folders,
-		repository: repo,
-	}, nil
+	parser     *resources.Parser
+	lister     resources.ResourceLister
+	folders    dynamic.ResourceInterface
+	dashboards dynamic.ResourceInterface
+
+	folderLookup map[string]string // folder path > ID that exists
+	changes      []ResourceFileChange
```

Comment:
I am not too convinced about this strategy. Having some many fields and specially modifying some of them on different functions makes very hard to follow and control. Some of those things are things ephemeral for which local variables are very suitable for (e.g. `changes`, `folderLookpup`). Other things are just something we want to return (e.g. `JobStatus`, `syncStatus` and `summary`).

Reply:
still pending to address / discuss